nuclear/roles/CMakeLists.txt

```diff
@@ -2,67 +2,66 @@
+# This gets all of our role files
+file(GLOB_RECURSE role_files "${role_source_dir}/*.role")
 
-# Backup the state of all the role variables before we re read them This is needed so we can delete roles that no longer
-# exist
+# Backup the state of all the role variables before we re read them. This is needed so we can delete roles that no
+# longer exist
 get_cmake_property(vars CACHE_VARIABLES)
 foreach(var ${vars})
-  if(var MATCHES "^ROLE_[a-zA-Z0-9]+")
+  if(var MATCHES "^ROLE_[a-zA-Z0-9_-]+")
     set("BACKUP_${var}" ${${var}})
     unset(${var} CACHE)
   endif()
 endforeach(var)
 
-# Find all of our role files
-foreach(rolefile ${rolefiles})
-  get_filename_component(roleext ${rolefile} EXT)
+# We need the NUClear role function here so it can be used in the role files
+include(NUClearRole.cmake)
 
-  # Check it is a .role file
-  if(roleext STREQUAL .role)
+# Clear our NUClear modules variable so we can repopulate it
+unset(NUCLEAR_MODULES CACHE)
 
-    # Roles may be within subfolders - get the subfolder and name
-    file(RELATIVE_PATH role_folder ${role_source_dir} ${rolefile})
-    cmake_path(REMOVE_EXTENSION role_folder)
-    string(REPLACE "/" "." role_folder ${role_folder})
+# Clear our NUClear roles variable so we can repopulate it
+unset(NUCLEAR_ROLES CACHE)
 
-    # If we had a backup as to if this role is enabled use it. Otherwise default to true.
-    if(DEFINED "BACKUP_ROLE_${role_folder}")
-      option("ROLE_${role_folder}" "Build the role ${role_folder}" ${BACKUP_ROLE_${role_folder}})
-    else()
-      option("ROLE_${role_folder}" "Build the role ${role_folder}" ON)
-    endif()
-  endif()
-endforeach(rolefile)
+# Find all of our role files and set our NUClear roles based on the ROLES cache variables that are set to true. This
+# makes it so that only the roles that are enabled will be built
+foreach(role_file ${role_files})
+  # Get path to role and the role name and add it to our list of roles
+  file(RELATIVE_PATH role_path ${role_source_dir} ${role_file})
+  get_filename_component(role_name ${role_path} NAME_WE)
 
-# Set our NUClear roles based on the ROLES cache variables that are set to true This makes it so that only the roles
-# that are enabled will be built
-get_cmake_property(vars CACHE_VARIABLES)
-unset(NUCLEAR_ROLES CACHE)
-foreach(var ${vars})
-  if(var MATCHES "^ROLE_[a-zA-Z0-9]+")
-    if(${var})
-      string(SUBSTRING ${var} 5 -1 role_folder)
-      string(REPLACE "." "/" role_folder ${role_folder})
-      set(NUCLEAR_ROLES
-          ${NUCLEAR_ROLES} ${role_folder}
-          CACHE INTERNAL "A list of the roles that are enabled to be built." FORCE
-      )
-    endif()
+  # Convert path to underscore separated folders
+  get_filename_component(role_path ${role_path} DIRECTORY)
+
+  # Join the role name with the role path again
+  if(role_path)
+    string(REPLACE "/" "_" role_folder ${role_path})
+    string(CONCAT role_folder ${role_folder} ";" ${role_name})
```

Comment:
Not sure `role_folder` is actually used anywhere? (either in this file or in `NUClearRole.cmake`)